Alco/manual table publishing

alco opened the pull request.

msfstef commented on packages/sync-service/config/runtime.exs:
```diff
@@ -225,7 +225,14 @@ config :electric,
   tcp_send_timeout:
-    env!("ELECTRIC_TCP_SEND_TIMEOUT", &Electric.Config.parse_human_readable_time!/1, nil)
+    env!("ELECTRIC_TCP_SEND_TIMEOUT", &Electric.Config.parse_human_readable_time!/1, nil),
+  manual_table_publishing?: env!("ELECTRIC_MANUAL_TABLE_PUBLISHING", :boolean, nil),
```
nit: could you add this in the docs in the website as part of this PR?

msfstef commented on packages/sync-service/lib/electric/replication/publication_manager.ex:
```diff
@@ -184,6 +188,8 @@ defmodule Electric.Replication.PublicationManager do
       db_pool: opts.db_pool,
       pg_version: opts.pg_version,
+      can_alter_publication?: Map.get(opts, :can_alter_publication?, true),
+      manual_table_publishing?: Map.get(opts, :manual_table_publishing?, false),
```
nit: do we need a fallback since we have defined a default in the nimble schema? seems redundant

msfstef commented on packages/sync-service/lib/electric/replication/publication_manager.ex:
```diff
@@ -285,6 +294,27 @@ defmodule Electric.Replication.PublicationManager do
              prepared_relation_filters: committed_filters
          }}
 
+      {:error, reason} when is_atom(reason) ->
```
this check feels a bit counterintuitive and hard to reason w.r.t. to the following cases - could we either 1. have a guard for the reason being one of the specific ones and 2. for the `manual_table_publishing?` and `can_alter_publication?` to be part of the guard for the case?

The reason being an atom feels like it could override the next two cases from occurring and without a clear enough reason

msfstef commented on packages/sync-service/lib/electric/replication/publication_manager.ex:
```diff
@@ -346,6 +377,34 @@ defmodule Electric.Replication.PublicationManager do
+      {:ok, state, []}
+    else
+      with {:ok, modified_relations} <-
+             Configuration.check_publication_for_missing_relations(
```
it feels counter intuitive that we call `update_publication` when we can't update the publication - which introduces the confusion in the cases discussed above - could we perhaps check these flags earlier on and call a `check_publication` function instead?

msfstef commented on packages/sync-service/test/electric/plug/low_privilege_router_test.exs:
```diff
@@ -0,0 +1,48 @@
+    setup [:with_complete_stack, :with_router]
+
+    @tag with_sql: ["INSERT INTO items VALUES (gen_random_uuid(), 'test value 1')"]
+    test "GET fails to create initial snapshot when Electric does not own the user table ", %{
```
very nice test setup! might be worth adding a few more here? (I see you've added a few integration tests but the router tests are easier to read/modify/run)

msfstef commented on packages/sync-service/lib/electric/postgres/configuration.ex:
```diff
@@ -11,6 +11,65 @@ defmodule Electric.Postgres.Configuration do
+          [Electric.oid_relation()],
+          filters()
+        ) :: :ok | {:error, :misconfigured_replica_identity | :table_missing_from_publication}
+  def check_publication_for_missing_relations(
```
nit: this implies that it is checking for missing relations, but it is actually checking that the given relations are present and configured correctly, so perhaps a `check_publication_validity` or `correctness` or `check_publication_configuration` to mirror the `configure_publication!` counterpart would be more clear?

msfstef reviewed (commented):
Very nice work - glad to see this approach working!

Left some comments specifically w.r.t. to the publication manager code being more explicit + adding the config to the website as part of this PR!

alco commented:
Thanks for the valuable feedback, @msfstef . I have addressed all of it.

alco commented on packages/sync-service/lib/electric/replication/publication_manager.ex:
```diff
@@ -251,39 +257,58 @@ defmodule Electric.Replication.PublicationManager do
+        {:ok, modified_relations} ->
+          update_relation_filters(state, modified_relations)
+
+        {:error, reason} ->
```
Brilliant observation!

alco commented on packages/sync-service/lib/electric/postgres/configuration.ex:
```diff
@@ -11,6 +11,76 @@ defmodule Electric.Postgres.Configuration do
+    if MapSet.size(diff_set) == 0 do
+      :ok
+    else
+      {:error, :tables_missing_from_publication}
```
Agreed. Fixing.

alco commented on packages/sync-service/lib/electric/shapes/api.ex:
```diff
@@ -545,13 +545,19 @@ defmodule Electric.Shapes.Api do
+      {:error, %SnapshotError{type: :missing_privilege} = error} ->
+        Logger.warning("Failed to create snapshot for #{shape_handle}: #{error.message}")
+        message = "Unable to create initial snapshot: " <> error.message
+        Response.error(request, message, status: 403)
```
Good point. I'm changing it to 503, known_error.

alco commented:
@icehaunter Thanks for the extremely useful insights! I have addressed your comments and added new commits to this PR. I'm not 100% sure about the correctness of the last one. I had to do that because otherwise the shape consumer would be the first to return a "must refetch" error to the snapshot waiter and the error from the publication manager would be ignored, breaking quite a large chunk of error reporting logic introduced in this PR.
